src/app/reporting/ReportScheduler.h

```diff
@@ -0,0 +1,160 @@
+        virtual void CancelTimer(void * context)                                       = 0;
+        virtual bool IsTimerActive(void * context)                                     = 0;
+        virtual Timestamp GetCurrentMonotonicTimestamp()                               = 0;
+    };
+
+    class ReadHandlerNode : public IntrusiveListNodeBase<>
+    {
+    public:
+        using TimerCompleteCallback = void (*)();
+
+        ReadHandlerNode(ReadHandler * aReadHandler, TimerDelegate * aTimerDelegate, TimerCompleteCallback aCallback) :
+            mTimerDelegate(aTimerDelegate), mCallback(aCallback)
+        {
+            VerifyOrDie(aReadHandler != nullptr);
+            VerifyOrDie(aTimerDelegate != nullptr);
+            VerifyOrDie(aCallback != nullptr);
+
+            mReadHandler = aReadHandler;
+            SetIntervalsTimeStamp(aReadHandler);
+        }
+        ReadHandler * GetReadHandler() const { return mReadHandler; }
+        /// @brief Check if the Node is reportable now, meaning its readhandler was made reportable by attribute dirtying and
+        /// handler state, and minimal time interval last reporthas elapsed since, or the maximal time interval since last report
+        /// has elapsed
+        bool IsReportableNow() const
+        {
+            // TODO: Add flags to allow for test to simulate waiting for the min interval or max intrval to elapse when integrating
+            // the scheduler in the ReadHandler
+            Timestamp now = mTimerDelegate->GetCurrentMonotonicTimestamp();
+            return (mReadHandler->IsGeneratingReports() &&
+                    ((now >= mMinTimestamp && mReadHandler->IsDirty()) || now >= mMaxTimestamp));
+        }
+
+        void SetIntervalsTimeStamp(ReadHandler * aReadHandler)
+        {
+            uint16_t minInterval, maxInterval;
+            aReadHandler->GetReportingIntervals(minInterval, maxInterval);
+            Timestamp now = mTimerDelegate->GetCurrentMonotonicTimestamp();
+            mMinTimestamp = now + System::Clock::Seconds16(minInterval);
+            mMaxTimestamp = now + System::Clock::Seconds16(maxInterval);
+        }
+
+        void RunCallback() { mCallback(); }
+
+        Timestamp GetMinTimestamp() const { return mMinTimestamp; }
+        Timestamp GetMaxTimestamp() const { return mMaxTimestamp; }
+
+    private:
+        TimerDelegate * mTimerDelegate;
+        TimerCompleteCallback mCallback;
+        ReadHandler * mReadHandler;
+        Timestamp mMinTimestamp;
+        Timestamp mMaxTimestamp;
+    };
+
+    ReportScheduler(TimerDelegate * aTimerDelegate) : mTimerDelegate(aTimerDelegate) {}
+    /**
+     *  Interface to act on changes in the ReadHandler reportability
+     */
+    virtual ~ReportScheduler() = default;
+
+    /// @brief Register a ReadHandler to the scheduler, will schedule report
+    /// @param aReadHandler read handler to register
+    virtual CHIP_ERROR RegisterReadHandler(ReadHandler * aReadHandler) = 0;
```

Comment:
Will do, the point was that technically, at the point where I tested TestReadHandlerList, I hadn't tested my Observer callbacks yet so I didn't want to rely on them. Since the Test is already a friend class of ReportSchedulerImpl, how about I move this to a private function existing only in ReportSchedulerImpl?

Reply:
> how about I move this to a private function existing only in ReportSchedulerImpl?

Sounds great, and then it can be non-virtual.  And then once we are done hooking everything up in ReadHandler we can re-evaluate which functions we want to have.